Restore OAuth support

ajdavis opened the pull request:
Partly addresses #135. This restores some OAuth logic that was lost in a large merge, and it updates the lost logic to be compatible with the latest wee-slack code.

The old logic had simply printed the Slack token to the buffer. I added a new feature: the new logic now automatically appends the token to the "plugins.var.python.slack.slack_api_token" setting and tells the user to restart weechat.

There are two more missing features after this which I propose to add:

1. Fix the OAuth redirect failure. Open a local webserver listening on a random available port, then include the `redirect_uri` parameter to redirect the auth flow back to localhost (if possible, this'll require research and maybe help from you to configure the wee-slack app on slack.com)
2. Update the `config_changed` callback to reload teams if the slack_api_token setting has changed

tfheen commented on wee_slack.py:
```diff
@@ -2964,6 +2964,50 @@ def me_command_cb(data, current_buffer, args):
+                w.prnt(current_buffer,
+                       "Success! Added team \"%s\"" % (d['team_name'],))
+                w.prnt(current_buffer,
+                       "Please restart weechat")
```
This should just ask them to reload wee-slack.

ajdavis replied:
Done, thanks.

ajdavis commented:
New patch, thanks for reviewing.

trygveaa commented on wee_slack.py:
```diff
@@ -2964,6 +2964,50 @@ def me_command_cb(data, current_buffer, args):
+
+
+"""
+        w.prnt(current_buffer, message)
```
I don't think we should add anything other than the actual messages to channel buffers, so I would want these messages to be printed to the team buffer instead. That goes for all the messages in this command.

ajdavis replied:
Done.

trygveaa commented on wee_slack.py:
```diff
@@ -2964,6 +2964,50 @@ def me_command_cb(data, current_buffer, args):
+                "https://slack.com/api/oauth.access?"
+                "client_id={}&client_secret={}&code={}"
+            ).format(CLIENT_ID, CLIENT_SECRET, aargs[1])
+            ret = urllib.urlopen(uri).read()
```
Hm, we go through weechat for all other requests (except for the websocket), but maybe it's fine to use urllib for this. It's certainly less hassle to use urllib.

trygveaa commented on wee_slack.py:
```diff
@@ -2964,6 +2964,50 @@ def me_command_cb(data, current_buffer, args):
+                w.prnt(current_buffer,
+                       "Please reload wee-slack")
+            else:
+                w.prnt(current_buffer, "Failed! Error is: " + d['error'])
```
Even though this is probably printed right after you issue the command, it's always better to give some more context in the error messages. So something like:

```
"ERROR: Couldn't get OAuth token: {}".format(d['error'])
```

ajdavis replied:
Done.

trygveaa commented on wee_slack.py:
```diff
@@ -3615,6 +3659,9 @@ def get_string(self, key):
         return int(w.config_get_plugin(key))
 
+    def is_default(self, key):
+        return w.config_get_plugin(key) == self.default_settings.get(key)
```
`w.config_get_plugin` returns a string while `default_settings.get` returns a `Settings` object, so this comparison doesn't work.

trygveaa commented on wee_slack.py:
```diff
@@ -2964,6 +2964,50 @@ def me_command_cb(data, current_buffer, args):
+                w.prnt(current_buffer,
+                       "Success! Added team \"%s\"" % (d['team_name'],))
+                w.prnt(current_buffer,
+                       "Please reload wee-slack")
```
I guess most people would know how to reload plugins, but we could always be helpful and tell them what to run:

```
"Please reload wee-slack with: /script reload slack"
```

ajdavis commented on wee_slack.py:
```diff
@@ -2964,6 +2964,50 @@ def me_command_cb(data, current_buffer, args):
 
+def command_register(data, current_buffer, args):
+    CLIENT_ID="2468770254.51917335286"
+    CLIENT_SECRET="dcb7fe380a000cba0cca3169a5fe8d70" #this is not really a secret
```
Done. (In my defense, I know this isn't PEP8, I'm cherry-picking an old commit by @rawdigits.)

ajdavis commented on wee_slack.py:
```diff
@@ -2964,6 +2964,50 @@ def me_command_cb(data, current_buffer, args):
+    CLIENT_SECRET="dcb7fe380a000cba0cca3169a5fe8d70" #this is not really a secret
+    if args == 'register':
+        message = """
+#### Retrieving a Slack token via OAUTH ####
```
Done.

ajdavis commented on wee_slack.py:
```diff
@@ -2964,6 +2964,50 @@ def me_command_cb(data, current_buffer, args):
+
+
+"""
+        w.prnt(current_buffer, message)
```
Done.

ajdavis commented:
I've fixed the ``is_default`` logic, nice catch. I also updated README. The instructions there are temporary, I'm going to make OAuth much simpler in my next patch.

ajdavis commented:
Hi! I think I've addressed all your comments. If you merge this I can proceed to the next steps, which will make the whole OAuth flow much smoother.

trygveaa commented on wee_slack.py:
```diff
@@ -2964,6 +2965,52 @@ def me_command_cb(data, current_buffer, args):
+            3) Click "Authorize" in the browser **IMPORTANT: the redirect will fail, this is expected**
+            4) Copy the "code" portion of the URL to your clipboard
+            5) Return to weechat and run `/slack register [code]`
+""")
```
This line can also be indented.

ajdavis replied:
Done.

trygveaa commented on wee_slack.py:
```diff
@@ -2964,6 +2965,52 @@ def me_command_cb(data, current_buffer, args):
+        return
+
+    aargs = args.split(None, 2)
+    if len(aargs) != 2:
```
Did you see my request for giving more useful names to the parts of args? I'd rather do it like this:

```
_, oauth_code, *extra_args = args.split()
if extra_args:
```

ajdavis replied:
Sorry, I think I missed that request. I like your idea but it's Python 3-only so I've modified it.

trygveaa replied:
No problem. And my bad, I thought it worked with python 2.

trygveaa commented on wee_slack.py:
```diff
@@ -2964,6 +2965,52 @@ def me_command_cb(data, current_buffer, args):
+
+    aargs = args.split(None, 2)
+    if len(aargs) != 2:
+        team.buffer_prnt("ERROR: invalid args to register")
```
I think this message would be clearer if it says: `ERROR: too many arguments given for register command`

ajdavis replied:
Done.

trygveaa commented on wee_slack.py:
```diff
@@ -2964,6 +2965,52 @@ def me_command_cb(data, current_buffer, args):
 
+def command_register(data, current_buffer, args):
+    e = EVENTROUTER
+    team = e.weechat_controller.buffers[current_buffer].team
```
My suggestion of printing to the team wasn't that good after all. If you aren't connected to a team already, this will fail.

This command doesn't really have any relevance to the possibly existing teams or where you run it though, so I think we should just always print to the core buffer for this command. That can be done with `w.prnt("", message)`.

ajdavis replied:
Done.

trygveaa commented on wee_slack.py:
```diff
@@ -2964,6 +2965,52 @@ def me_command_cb(data, current_buffer, args):
+                            ','.join([tok, d['access_token']]))
+
+    team.buffer_prnt("Success! Added team \"%s\"" % (d['team_name'],))
+    team.buffer_prnt("Please reload wee-slack")
```
Can you change this message to: `Please reload wee-slack with: /script reload slack`?

ajdavis replied:
Done.

trygveaa commented on README.md:
````diff
@@ -88,10 +88,35 @@ weechat
+/slack register
+```
+
+This command opens your browser and prompts you to authorize WeeChat with Slack.
````
Well, it doesn't open the browser. It just prints a link, so maybe: `This command prints a link you should open in your browser [...]`

ajdavis replied:
Done.

ajdavis commented:
New patch, thanks for your review.

oblitum commented:
@ajdavis thanks for working on this! It's my stumbling block to use Slack on WeeChat. Do you know how long the token stays valid, is the authorization permanent?

trygveaa commented:
@oblitum: The token is valid permanently, unless you revoke it. From https://api.slack.com/docs/oauth
> The token will continue functioning until the installing user either revokes the token and/or uninstalls your application.
> Please note that these access tokens do not expire.

oblitum commented:
@trygveaa OK, thanks.

trygveaa commented:
> Fix the OAuth redirect failure. Open a local webserver listening on a random available port

@ajdavis: If you plan on doing this, note that not everyone runs weechat on their local machine (it's common to open it through ssh), so localhost for wee-slack may not be where the user runs their browser.

trygveaa commented:
Thanks!

oblitum commented:
Great! but also sad that I didn't realize apps would need to be authorized peer team.... sigh.
